docs/advanced-features/good-practice.md

```diff
@@ -101,5 +104,772 @@ If enabled (see set-up instructions above), these events will be
+|`2023-06-14 11:33:42.000`|`Users.GrantAdmin`|`Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/114.0`|*REDACTED*|`{"username":"*REDACTED*"}`||`{}`||
+|`2023-06-14 11:34:16.000`|`Roles.Update`|`Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/114.0`|*REDACTED*|`{"input":{"name":"*REDACTED*","policies":[*REDACTED*]},"managed":true,"id":"*REDACTED"}`|`{"role":{*REDACTED*},"__typename":"RoleUpdateSuccess"}`|`{"graphql_path":"roleUpdate"}`||
+|`2023-06-14 11:34:26.000`|`Auth.Logout`|`Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/114.0`|*REDACTED*|`{}`||`{}`||
+
+Which users were active this month and what actions they performed?
+
+```sql
+SELECT
+json_extract_scalar(useridentity, '$.id') as userid,
+array_agg(DISTINCT json_extract_scalar(useridentity, '$.username')) as usernames,
+array_agg(DISTINCT json_extract_scalar(useridentity, '$.email')) as emails,
+array_agg(DISTINCT json_extract_scalar(useridentity, '$.isadmin')) as isadmin_values,
+array_agg(DISTINCT json_extract_scalar(useridentity, '$.roleid')) as roles,
+array_agg(DISTINCT sourceipaddress) as ips,
+min(eventtime) as time_first,
+max(eventtime) as time_last,
+array_agg(DISTINCT eventname) as actions
+FROM audit_trail
+WHERE date BETWEEN date_format(current_date, '%Y/%m/01') AND date_format(current_date, '%Y/%m/31')
+AND json_extract_scalar(useridentity, '$.type') = 'QuiltUser'
+GROUP BY json_extract_scalar(useridentity, '$.id')
+```
+
+Example query result:
+
+|`userid`|`usernames`|`emails`|`isadmin_values`|`roles`|`ips`|`time_first`|`time_last`|`actions`|
+|--------|-----------|--------|----------------|-------|-----|------------|-----------|---------|
+|*REDACTED*|`[admin]`|`[admin@example.com]`|`[true]`|`[*REDACTED*]`|`[*REDACTED*]`|`2023-06-08 00:16:48.000`|`2023-06-14 11:34:26.000`|`[Users.Disable, Users.Enable, Users.GrantAdmin, Roles.Update, Auth.Logout, Users.RevokeAdmin, Auth.GetAWSCredentials, Auth.Login, Users.List]`|
+|*REDACTED*|`[user]`|`[user@example.com]`|`[true, false]`|`[*REDACTED*]`|`[*REDACTED*]`|`2023-06-08 00:16:26.000`|`2023-06-14 13:10:02.000`|`[Users.List, Auth.GetAWSCredentials]`|
+|*REDACTED*|`[_canary]`|`[canary@quiltdata.io]`|`[false]`|`[*REDACTED*]`|`[*REDACTED*]`|`2023-06-08 00:37:21.000`|`2023-06-14 12:38:20.000`|`[Auth.GetAWSCredentials, Auth.ServiceLogin]`|
+
+Which unique users logged in this month and which role did they have and were
+they an admin?
+
+```sql
+SELECT
+json_extract_scalar(useridentity, '$.id') as userid,
+array_agg(DISTINCT json_extract_scalar(useridentity, '$.username')) as usernames,
+array_agg(DISTINCT json_extract_scalar(useridentity, '$.email')) as emails,
+array_agg(DISTINCT json_extract_scalar(useridentity, '$.isadmin')) as isadmin_values,
+array_agg(DISTINCT json_extract_scalar(useridentity, '$.roleid')) as roles
+FROM audit_trail
+WHERE date BETWEEN date_format(current_date, '%Y/%m/01') AND date_format(current_date, '%Y/%m/31')
```

Comment:
`date` is some kind of keyword in SQL (at least for the syntax highlighter) so maybe quote it if it's a column name.

Reply:
well, it works correctly in athena, so...